### PR TITLE
Bug 1732164: - [metrics] Report metrics for default operator-sources

### PR DESCRIPTION
**Pull request by varshaprasad96**
Problem:
The OpenShift Marketplace does not have a method to communicate to the user,
when a default operator source fails.
Solution:
This PR adds the custom metrics collection logic for three major operator sources -
redhat, community and certified. When an operator source fails to reconcile,
the respective operator-source name, error message and count of failures
are reported by the prometheus. Alerts are created on these metrics, so as to 
integrate them with telemetry server.

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -10,6 +10,10 @@ required = [
   "sigs.k8s.io/controller-tools/pkg/crd/generator",
```
Comment (awgreene) [inline Gopkg.toml]: Please follow the 50/72 rule in your commit.

File: cmd/manager/main.go [hunk 2]
```diff
@@ -26,10 +32,13 @@ import (
 const (
 	// TODO: resyncInterval is hardcoded to 1 hour now, it would have to be
 	// configurable on a per OperatorSource level.
-	resyncInterval = time.Duration(60) * time.Minute
-
+	resyncInterval             = time.Duration(60) * time.Minute
 	initialWait                = time.Duration(1) * time.Minute
 	updateNotificationSendWait = time.Duration(10) * time.Minute
+
+	// Endpoint and path where the metrics would be exposed.
```
Comment (awgreene) [inline cmd/manager/main.go]: This should be `// Endpoint and path where the metrics are be exposed.`

File: cmd/manager/main.go [hunk 3]
```diff
@@ -57,10 +69,15 @@ func main() {
 		log.Fatal(err)
 	}
 
-	// Create a new Cmd to provide shared dependencies and start components
-	mgr, err := manager.New(cfg, manager.Options{Namespace: namespace})
+	// Create a new Cmd to provide shared dependencies and start components at the specified
```
Comment (awgreene) [inline cmd/manager/main.go]: `... in the specified namespace.`

File: cmd/manager/main.go [hunk 4]
```diff
@@ -83,6 +100,20 @@ func main() {
...
 
+	ctx := context.TODO()
+
+	// Create Service object to expose the metrics port.
+	service, err := metrics.ExposeMetricsPort(ctx, metricsPort)
+	if err != nil {
+		log.Errorf("Unable to expose metrics port: %v", err)
+	}
+
+	// Create a serviceMonitor for the namespace, and register the service created above to it.
```
Comment (awgreene) [inline cmd/manager/main.go]: Please remove the `,`

File: cmd/manager/main.go [hunk 5]
```diff
@@ -83,6 +100,20 @@ func main() {
 		exit(err)
 	}
 
+	ctx := context.TODO()
+
+	// Create Service object to expose the metrics port.
```
Comment (awgreene) [inline cmd/manager/main.go]: `Create a Service...`

File: pkg/customMetrics/customMetrics.go [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Registry-namespaces of operator sources.
+const (
```
Comment (awgreene) [inline pkg/customMetrics/customMetrics.go]: This will need to be updated based on our offline conversation with @aravindhp

File: pkg/customMetrics/customMetrics.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	REDHAT_OPERATOR_NS    = "redhat-operators"
```
Comment (awgreene) [inline pkg/customMetrics/customMetrics.go]: Please add comments on all these constants.

File: pkg/customMetrics/customMetrics.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+	// List of custom metrics which are to be collected.
+	metricsList = []prometheus.Collector{
+		redhatErrorCount,
+		communityErrorCount,
+		certifiedErrorCount,
+		customOperatorErrorCount,
+	}
+
+	// custom metrics port and path where the metrics would be
```
Comment (awgreene) [inline pkg/customMetrics/customMetrics.go]: Same as above... Why is this described in two places?

File: pkg/customMetrics/customMetrics.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+	// exposed.
+	metricsPath = "/metrics"
+	metricsPort = ":8383"
```
Comment (awgreene) [inline pkg/customMetrics/customMetrics.go]: Same as above... Why is this described in two places? You could probably make the `metricsPath, metricsPort, and metricsHost` global constants within the metrics package.

File: pkg/customMetrics/customMetrics.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
+package customMetrics
```
Comment (awgreene) [inline pkg/customMetrics/customMetrics.go]: I vote to rename this package metrics.

File: pkg/operatorsource/handler.go [hunk 11]
```diff
@@ -4,15 +4,23 @@ import (
...
 	"github.com/operator-framework/operator-marketplace/pkg/appregistry"
+	"github.com/operator-framework/operator-marketplace/pkg/customMetrics"
 	"github.com/operator-framework/operator-marketplace/pkg/datastore"
 	"github.com/operator-framework/operator-marketplace/pkg/phase"
 	log "github.com/sirupsen/logrus"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
 )
 
+// Names of the operator-sources shipped with marketplace
```
Comment (awgreene) [inline pkg/operatorsource/handler.go]: There is no need to redefine these constants.

File: pkg/operatorsource/handler.go [hunk 12]
```diff
@@ -123,3 +137,18 @@ func (h *operatorsourcehandler) transition(ctx context.Context, logger *log.Entr
 
 	return reconciliationErr
 }
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func updateMetrics(opsrc string) {
```
Comment (awgreene) [inline pkg/operatorsource/handler.go]: I would move this to the `metrics` package.

File: pkg/customMetrics/customMetrics.go (hunk 10 above)
Comment (varshaprasad96) [reply pkg/customMetrics/customMetrics.go]: The default package provided by SDK for metrics collection is also named as `metrics`.  Just for the differentiation, this is named customMetrics. If that's fine,  shall I revert back to metrics?

Comment (awgreene) [reply pkg/customMetrics/customMetrics.go]: That is fine with me.

File: cmd/manager/main.go [hunk 13]
```diff
@@ -57,10 +69,15 @@ func main() {
...
-	mgr, err := manager.New(cfg, manager.Options{Namespace: namespace})
+	// Create a new Cmd to provide shared dependencies and start components at the specified
+	// namespace.
+	mgr, err := manager.New(cfg, manager.Options{
+		Namespace:      namespace,
+		MapperProvider: restmapper.NewDynamicRESTMapper,
+	})
 	if err != nil {
 		log.Fatal(err)
+		os.Exit(1)
```
Comment (galletti94) [inline cmd/manager/main.go]: why are we using `os.Exit(1)` here?

File: pkg/operatorsource/configuring.go [hunk 14]
```diff
@@ -5,7 +5,7 @@ import (
 	"errors"
 
 	"github.com/operator-framework/operator-marketplace/pkg/apis/operators/shared"
-	"github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
+	v1 "github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
```
Comment (galletti94) [inline pkg/operatorsource/configuring.go]: Please check that you are using `gofmt` as your  formatting tool in your editor

File: pkg/operatorsource/handler.go [hunk 15]
```diff
@@ -4,15 +4,23 @@ import (
 	"context"
 
 	"github.com/operator-framework/operator-marketplace/pkg/apis/operators/shared"
-	"github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
+	v1 "github.com/operator-framework/operator-marketplace/pkg/apis/operators/v1"
```
Comment (galletti94) [inline pkg/operatorsource/handler.go]: same comment as above

File: pkg/operatorsource/handler.go [hunk 16]
```diff
@@ -123,3 +137,18 @@ func (h *operatorsourcehandler) transition(ctx context.Context, logger *log.Entr
 
 	return reconciliationErr
 }
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func updateMetrics(opsrc string) {
+	switch opsrcName := opsrc; opsrcName {
+	case REDHAT_OPERATOR:
+		customMetrics.CounterForRedHatOpsrcError(opsrcName).Inc()
```
Comment (galletti94) [inline pkg/operatorsource/handler.go]: Is the implementation of `Inc()` handled by prometheus? I am just curious because usually the first example of distributed programming errors is one of incrementing counters lol

File: cmd/manager/main.go (hunk 13 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: Sorry, this wouldn't be needed. I will rebase and push the changes.

File: pkg/operatorsource/handler.go (hunk 16 above)
Comment (varshaprasad96) [reply pkg/operatorsource/handler.go]: True!. Fortunately, Prometheus is handling that here.

File: pkg/metrics/metrics.go [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+import (
+	"net/http"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+const (
+	// Registry-namespaces of operator sources.
+	REDHAT_OPERATOR_NS    = "redhat-operators"
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: GO Variables should not be in caps. Applies everywhere.

File: pkg/metrics/metrics.go [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+	COMMUNITY_OPERATOR_NS = "community-operators"
+	CERTIFIED_OPERATOR_NS = "certified-operators"
+
+	// Default values which are to be used as labels
+	// for prometheus metrics.
+	REGISTRY_NAMESPACE = "registry_ns"
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: What is `REGISTRY_NAMESPACE` used for?

File: pkg/metrics/metrics.go [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+var (
+	redhatErrorCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "redhat_operator_reconcile_failure_count",
+			Help: "Monotonic count of times redhat-opsrc was unable to reconcile",
+		},
+		[]string{REGISTRY_NAMESPACE, OPSRC_NAME},
+	)
+	communityErrorCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: Are we able to expose a string as an error? Knowing how many times an operator failed is not helpful, knowing why it failed is.

File: cmd/manager/main.go [hunk 20]
```diff
@@ -85,6 +98,27 @@ func main() {
 		exit(err)
 	}
 
+	ctx := context.TODO()
+
+	// convert the value of MetricsPort from string to int32.
+	port, err := strconv.ParseInt(customMetrics.MetricsPort, 10, 32)
```
[the remark this answers is not in the document]
Comment (varshaprasad96) [reply cmd/manager/main.go]: Its because the metricsPort in `http.ListenAndServe` requires a String input. So, conversion is need at anyone of the place.

File: pkg/metrics/metrics.go (hunk 18 above)
Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: It is a label used, for Prometheus metrics.

**Comment by varshaprasad96** [comment]
The PR has been updated to handle two cases for reporting errors of default operators:
1. Marketplace cannot reach Quay, as it is unavailable ("Status (404) error")
2. The provided host does not exists (No such host exists)
Currently, for other cases, a generic error message is returned.

File: pkg/metrics/metrics.go [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+// RecordMetrics starts the server, records the custom metrics and exposes
+// the metrics at the specified endpoint.
+func RecordMetrics() {
+	// Register metrics for the operator with the prometheus.
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	http.Handle(METRICSPATH, promhttp.Handler())
+	port := ":" + METRICSPORT
+	go http.ListenAndServe(port, nil)
```
Comment (ecordell) [inline pkg/metrics/metrics.go]: This is registering and serving prom from the global muxer - I assume there's a health check also on it somewhere?

That would mean the health check is exposed on two ports, which probably won't cause an error but is kind of weird.

File: pkg/metrics/metrics.go [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+	ERROR      = "opsrc_error"
+	OPSRC_NAME = "opsrc_name"
+
+	// Path, Port and Host where custom metrics would be exposed.
+	METRICSPATH = "/metrics"
+	METRICSPORT = "8383"
+	METRICSHOST = "0.0.0.0"
+
+	// Error message to be returned when quay.io is down.
+	UNREACHABLE_ERROR = "Quay.io is unavailabe"
```
Comment (ecordell) [inline pkg/metrics/metrics.go]: it's a lot easier to query prometheus when label values don't have spaces. What about something like `quay_unavailable`, `no_host`, etc?

File: pkg/metrics/metrics.go [hunk 23]
```diff
@@ -0,0 +1,90 @@
...
+
+// counterForOpsrcError returns the number of times error was encountered while
+// contacting quay for operstor-sources.
+func counterForOpsrcError(opsrc, opsrcError string) prometheus.Counter {
+	return operatorSourceError.WithLabelValues(opsrc, opsrcError)
+}
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func UpdateMetrics(opsrc, opsrcError string) {
```
Comment (ecordell) [inline pkg/metrics/metrics.go]: nit: to make it clear that that you should not pass any old value here in the future, it might make sense to pass `opsrcError` as an enum value instead of an arbitrary string.

File: pkg/metrics/metrics.go (hunk 21 above)
Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: Yes, these are the same port and path values which SDK uses to expose metrics and health check occurs here too. 
Shall I change, or is it fine to use it from a global mux?

File: pkg/metrics/metrics.go (hunk 23 above)
Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: Actually, I am taking `opsrcError` based on the error message we get from `ListPackages` during Configuring phase. So it can be either of the three values (no_host, quay_down, or a custom error message) as of now. Did you mean that we pass the label values as constants ?

File: pkg/metrics/metrics.go (hunk 21 above)
Comment (ecordell) [reply pkg/metrics/metrics.go]: Ah - if they're both configured on the same port that's fine. I should've checked 😄 I assumed they were configured on different ports.

File: pkg/metrics/metrics.go (hunk 23 above)
Comment (ecordell) [reply pkg/metrics/metrics.go]: I meant something along the lines of 

```go
type OpSrcError string

const (
        NO_HOST_ERROR OpSrcError = "no_host"
        UNREACHABLE_ERROR OpSrcError = "quay_unavailable"
)

func UpdateMetrics(opsrc, opSrcErr OpSrcError) {
  // ...
}
```

That way it's super clear to anyone calling `UpdateMetrics` that you shouldn't make up your own string and pass it in.

Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: Thanks @ecordell. Got it..

File: manifests/07_operator.yaml [hunk 24]
```diff
@@ -41,7 +41,6 @@ spec:
           - marketplace-operator
           args:
           - -registryServerImage=quay.io/openshift/origin-operator-registry
-          - -defaultsDir=/defaults
```
Comment (awgreene) [inline manifests/07_operator.yaml]: Did you mean to remove this line?

File: manifests/07_operator.yaml [hunk 25]
```diff
@@ -56,12 +55,12 @@ spec:
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.namespace
+            - name: OPERATOR_NAME
```
Comment (awgreene) [inline manifests/07_operator.yaml]: Why are you swapping the location of this env var?

File: cmd/manager/main.go [hunk 26]
```diff
@@ -118,6 +129,25 @@ func main() {
 		exit(err)
 	}
 
+	// convert the value of MetricsPort from string to int32.
+	port, err := strconv.ParseInt(customMetrics.METRICSPORT, 10, 32)
```
Comment (awgreene) [inline cmd/manager/main.go]: I don't believe that this will compile.

File: cmd/manager/main.go [hunk 27]
```diff
@@ -51,7 +58,7 @@ func printVersion() {
 func main() {
 	printVersion()
 
-	// Parse the command line arguments
+	// Parse the command line arguments for the registry server image
```
Comment (awgreene) [inline cmd/manager/main.go]: Please undo this change.

File: cmd/manager/main.go [hunk 28]
```diff
@@ -118,6 +129,9 @@ func main() {
 		exit(err)
 	}
 
+	// generate metrics and publish it at an endpoint.
```
Comment (awgreene) [inline cmd/manager/main.go]: Please capitalize the first letter in you comments, comments are sentences.

File: cmd/manager/main.go [hunk 29]
```diff
@@ -210,3 +224,28 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
 
 	return err
 }
+
+// GenerateMetrics creates a service, service monitor, starts a metrics server
+// and records the metrics to expose at the specified endpoint.
+
```
Comment (awgreene) [inline cmd/manager/main.go]: Please remove this newline.

File: cmd/manager/main.go [hunk 30]
```diff
@@ -210,3 +224,28 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
...
+func GenerateMetrics(cfg *rest.Config, namespace string) {
+
+	// Create a Service object to expose the metrics port.
+	service, err := metrics.ExposeMetricsPort(context.TODO(), metricsPort)
+	if err != nil {
+		log.Errorf("Unable to expose metrics port: %v", err)
+	}
+
+	// Ensure that an empty service object is not passed to create a service monitor.
```
Comment (awgreene) [inline cmd/manager/main.go]: `// Ensure that a nil pointer is not added to the slice`

File: cmd/manager/main.go [hunk 31]
```diff
@@ -3,10 +3,10 @@ package main
 import (
```
Comment (awgreene) [inline cmd/manager/main.go]: Given that this work is now tracked via a bug, we will need to rewrite the commit message and change the PR header.

File: cmd/manager/main.go [hunk 32]
```diff
@@ -93,9 +90,8 @@ func main() {
 	// watch for CatalogSources in targetNamespaces being deleted and recreate
 	// them.
 	mgr, err := manager.New(cfg, manager.Options{
-		Namespace:          namespace,
-		MapperProvider:     restmapper.NewDynamicRESTMapper,
-		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
```
Comment (awgreene) [inline cmd/manager/main.go]: Can you explain why this is removed? I assume it is to prevent us from reporting sdk metrics, but I want to confirm.

File: cmd/manager/main.go [hunk 33]
```diff
@@ -230,6 +226,13 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
 
 func GenerateMetrics(cfg *rest.Config, namespace string) {
 
+	// convert the value of MetricsPort from string to int32.
```
Comment (awgreene) [inline cmd/manager/main.go]: Please capitalize `// Convert`.

File: cmd/manager/main.go [hunk 34]
```diff
@@ -230,6 +226,13 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
...
+	port, err := strconv.ParseInt(customMetrics.MetricsPort, 10, 32)
+	if err != nil {
+		log.Errorf("Unable to parse metrics port: %v", err)
```
Comment (awgreene) [inline cmd/manager/main.go]: If this fails, won't there be an issue when you call `metricsPort := int32(port)`?

File: cmd/manager/main.go [hunk 35]
```diff
@@ -230,6 +226,13 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
 
 func GenerateMetrics(cfg *rest.Config, namespace string) {
 
+	// convert the value of MetricsPort from string to int32.
+	port, err := strconv.ParseInt(customMetrics.MetricsPort, 10, 32)
```
Comment (awgreene) [inline cmd/manager/main.go]: Why not just defined `customMetrics.MetricsPort` as an `int32`?

File: pkg/metrics/metrics.go [hunk 36]
```diff
@@ -0,0 +1,98 @@
...
+
+// RecordMetrics starts the server, records the custom metrics and exposes
+// the metrics at the specified endpoint.
+func RecordMetrics() {
+	// Register metrics for the operator with the prometheus.
+	registerMetrics()
+
+	// Start the server and expose the registered metrics.
+	http.Handle(MetricsPath, promhttp.Handler())
+	port := ":" + MetricsPort
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: If MetricsPort was a string, you could set port to `fmt.Sprintf(":%d", MetricsPort)`.

File: pkg/metrics/metrics.go [hunk 37]
```diff
@@ -0,0 +1,98 @@
...
+
+const (
+	// Default values which are to be used as labels
+	// for prometheus metrics.
+	Error     = "opsrc_error"
+	OpsrcName = "opsrc_name"
+
+	// Path, Port and Host where custom metrics would be exposed.
+	MetricsPath = "/metrics"
+	MetricsPort = "8383"
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: This should just be an `integer`

File: pkg/metrics/metrics.go [hunk 38]
```diff
@@ -0,0 +1,98 @@
...
+	MetricsHost = "0.0.0.0"
+
+	// Error message to be returned when quay.io is down.
+	UnreachableError OpsrcError = "quay_unavailabe"
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: Documentation for Global variables should start with the name of the global variable. This applies in all places. Please see marketplace code for examples.

File: pkg/metrics/metrics.go [hunk 39]
```diff
@@ -0,0 +1,98 @@
...
+	// Error message to be returned when host address is wrong.
+	NoHostExsistsError OpsrcError = "no_host_exsists"
+)
+
+// Metric variables that report error based on the operator-source
+// which fails to reconcile.
+var (
+	operatorSourceError = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "operator_source_failure_count",
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: This could be `opsrc_failures`, unless promethes has some convention I am unaware of. Thoughts @kevinrizza ?

Comment (awgreene) [reply pkg/metrics/metrics.go]: `opsrc_errors` works here as well.

File: pkg/metrics/metrics.go [hunk 40]
```diff
@@ -0,0 +1,98 @@
...
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func UpdateMetrics(opsrc, opsrcError string) {
+	counterForOpsrcError(opsrc, opsrcError).Inc()
+}
+
+// ErrorMessage takes the error which appears while Listing the packages,
+// parses it, and returns the error message which is to be published.
+func ErrorMessage(err string) OpsrcError {
```
Comment (awgreene) [inline pkg/metrics/metrics.go]: This is a good start, we will need to review how we wish to ID quay errors.

File: pkg/operatorsource/handler.go (hunk 15 above)
Comment (awgreene) [reply pkg/operatorsource/handler.go]: This still needs to be addressed in all places.

File: cmd/manager/main.go (hunk 32 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: Yes, `MetricsBindAddress` would run the command to register default metrics with at the port/path provided by SDK. Since we are adding our own custom metrics to it and registering them, this needs to be removed. Or else, we would be registering metrics to the same port twice, which will result in an error.

File: cmd/manager/main.go (hunk 34 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: Actually no. It would return a zero when anything other than a valid integer is passed (like a string).

File: pkg/metrics/metrics.go (hunk 37 above)
Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: MetricsPort as an integer cannot to passed to `http.ListenAndServe`. Casting is required in either one of the place (before starting the server, or before passing it to `ExposeMetrics` function)

File: cmd/manager/main.go (hunk 35 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: Because `http.ListenAndServe` requires a string.

File: pkg/metrics/metrics.go (hunk 39 above)
Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: Prometheus requirements are just to have "_" between words. Since this is of the type `counter` in Prometheus metrics, I named it as `operator_source_failure_count`. This could be anything though..

Comment (awgreene) [reply pkg/metrics/metrics.go]: That's good to know, let's see what name @kevinrizza thinks is good, perhaps `opsrc_failure_count`?

File: pkg/metrics/metrics.go [hunk 41]
```diff
@@ -0,0 +1,100 @@
...
+
+// counterForOpsrcError returns the number of times error was encountered while
+// contacting quay for operstor-sources.
+func counterForOpsrcError(opsrc, opsrcError string) prometheus.Counter {
+	return operatorSourceError.WithLabelValues(opsrc, opsrcError)
+}
+
+// updateMetrics takes the name of the operator source which fails to reconcile, and
+// increments the error count of the respective operator metric.
+func UpdateMetrics(opsrc, opsrcError string) {
```
Comment (ecordell) [inline pkg/metrics/metrics.go]: I think this should be `func UpdateMetrics(opsrc, opsrcError OpsrcError)`?

Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: The labels of the prometheus metrics only take in String type. So I created a function `OpsrcErrorToString`, which would just cast it into a string only while updating the labels.

Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: I had used that in `configuring.go`, while incrementing metrics. But that doesn't serve the purpose. So I have updated it here! :+1:

File: cmd/manager/main.go [hunk 42]
```diff
@@ -210,3 +219,30 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
...
+
+func generateMetrics(cfg *rest.Config, namespace string) {
+
+	// Create a Service object to expose the metrics port.
+	service, err := metrics.ExposeMetricsPort(context.TODO(), customMetrics.MetricsPort)
+	if err != nil {
+		log.Errorf("Unable to expose metrics port: %v", err)
+	}
+
+	// Ensure that a nil pointer is not added to the slice.
```
Comment (kevinrizza) [inline cmd/manager/main.go]: Generally I think adding a comment about the mechanics of the code is bad practice unless it is particularly unorthodox. I think in this case the purpose of checking for nil before generating the slice makes sense. If this is generally the only purpose, I think removing this comment makes sense here.


That being said if `service == nil` isn't it possible that we want to do something? Return an error perhaps? What state are we in if that happens? I think the same comment could be made just further above if `metrics.ExposeMetricsPort()` returns an error.

File: cmd/manager/main.go [hunk 43]
```diff
@@ -210,3 +219,30 @@ func ensureDefaults(cfg *rest.Config, scheme *kruntime.Scheme) error {
...
+	services := []*v1.Service{}
+	if service != nil {
+		services = append(services, service)
+	}
+
+	// Create a serviceMonitor for the namespace and register the service created above to it.
+	_, err = metrics.CreateServiceMonitors(cfg, namespace, services)
+	if err != nil {
+		log.Errorf("Error Creating Service Monitor: %v", err)
```
Comment (kevinrizza) [inline cmd/manager/main.go]: Same comment as above with respect to error handling. Is just logging appropriate?

File: deploy/role.yaml [hunk 44]
```diff
@@ -120,3 +121,12 @@ rules:
   - create
   - delete
   - update
+  - list
+  - watch
```
Comment (kevinrizza) [inline deploy/role.yaml]: This seems like the wrong thing to do. Why do we need to list and watch secrets for this? Shouldn't we just add another rbac grouping for pods if that is what's needed?

File: pkg/metrics/metrics.go [hunk 45]
```diff
@@ -0,0 +1,100 @@
...
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// Error to be  returned when marketplace is unable to reach quay.io.
```
Comment (kevinrizza) [inline pkg/metrics/metrics.go]: We should avoid specifically referencing quay.io whenever possible. Keep in mind that the fact that quay hosts app registry doesn't mean that it's the only way to get an app registry instance.

File: pkg/metrics/metrics.go (hunk 39 above)
Comment (kevinrizza) [reply pkg/metrics/metrics.go]: I think keeping our variable names short, especially when they are going to be displayed in a UI is good. `opsrc_failure_count` seems fine.

File: cmd/manager/main.go [hunk 46]
```diff
@@ -118,6 +124,9 @@ func main() {
 		exit(err)
 	}
 
+	// Generate metrics and publish it at an endpoint.
```
Comment (kevinrizza) [inline cmd/manager/main.go]: Should this be async? Though I suppose we care that we start recording before we start handling events so we don't miss any? Have we considered what all this initialization will do (in terms of time cost) if we don't at least serve the health check first?

File: pkg/metrics/metrics.go (hunk 40 above)
Comment (kevinrizza) [reply pkg/metrics/metrics.go]: Agreed. It seems like string comparison will be brittle (if these error messages change, this code will silently break).

Perhaps we can derive meaning from control flow instead? Or at least http error codes if we can get them from the client?

Comment (varshaprasad96) [reply pkg/metrics/metrics.go]: Yeah, true. We do get a http 404 error code when quay is not available. But when the endpoint is wrong, it remains in `configuring` phase, trying to connect. There is no specific error code returned in the error message. I am not sure on how to proceed with that case though..

File: cmd/manager/main.go [hunk 47]
```diff
@@ -36,8 +40,7 @@ import (
 const (
 	// TODO: resyncInterval is hardcoded to 1 hour now, it would have to be
 	// configurable on a per OperatorSource level.
-	resyncInterval = time.Duration(60) * time.Minute
-
+	resyncInterval             = time.Duration(60) * time.Minute
```
[the remark this answers is not in the document]
Comment (varshaprasad96) [reply cmd/manager/main.go]: Its due to formatting issues. Ill remove the spaces.

File: pkg/metrics/metrics.go (hunk 40 above)
Comment (kevinrizza) [reply pkg/metrics/metrics.go]: We put ourselves back in the configuring phase because we know there was some error and we want to retry. We should be able to dig into where we decide to put ourselves back into that phase and probably learn more. And, if there isn't something more specific being returned, is it possible for us to adjust how the appr client is giving a result? Are we really restricted to just strings? Could we update the error that is returned to have an `ErrorCode` field or something similar?

File: cmd/manager/main.go (hunk 46 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: The `RecordMetrics()` function uses the endpoint defined by SDK. Hence the health check does happen here before the metrics are served. The scrapping of data by Prometheus is synchronous. But yes, as you said we can decouple recording metrics logic from this function and start collecting metrics initially. Creation of service and service monitor could happen later.

File: cmd/manager/main.go (hunk 42 above)
Comment (varshaprasad96) [reply cmd/manager/main.go]: If the service is nil, then we return an error stating `Unable to expose metrics port`. That part  is just to make sure that we do not pass a nil to the next function. Because if `nil` value is passed to `CreateServiceMonitor`, then a segfault occurs. So, when a service is not created, then servicemonitor is not created either and we do not expose metrics.

Comment (awgreene) [reply cmd/manager/main.go]: I believe @kevinrizza is trying to point out that we may want to fail the operator rather than just logging the error and carrying on.

File: deploy/role.yaml (hunk 44 above)
Comment (varshaprasad96) [reply deploy/role.yaml]: `watch` is not needed. I just reconfirmed with the SDK team too. `list` would be required and is sufficient!

**Comment by varshaprasad96** [comment]
@awgreene  @ecordell - I have updated the code based on the suggestions provided. Please have a look at it.

**Comment by awgreene** [comment]
Closed in favor of #246
